autometa/common/external/prodigal.py

```diff
@@ -67,108 +155,42 @@ def run(assembly, nucls_out, prots_out, force=False, cpus=0, parallel=True):
-        tmpdir = os.path.join(outdir, 'tmp')
-        if not os.path.exists(tmpdir):
-            os.makedirs(tmpdir)
-        tmpnucl = '.'.join([outprefix, '{#}', 'fna'])
-        tmpprot = '.'.join([outprefix, '{#}', 'faa'])
-        tmpnucl_fpath = os.path.join(tmpdir, tmpnucl)
-        tmpprot_fpath = os.path.join(tmpdir, tmpprot)
-        jobs = f'-j{cpus}'
-        cmd = [
-            'parallel',
-            '--retries', '4',
-            '--joblog', log,
-            jobs,
-            '--pipe',
-            '--recstart',
-            '\'>\'',
-            '--linebuffer',
-            'prodigal',
-            '-a', tmpprot_fpath,
-            '-d', tmpnucl_fpath,
-            '-q',
-            '-p', 'meta',
-            '-m',
-            '-o', os.devnull,
-            '<', assembly,
-            '2>', os.devnull,
-        ]
+        annotate_parallel(
+            assembly=assembly, prots_out=prots_out, nucls_out=nucls_out, cpus=cpus
+        )
     else:
-        cmd = [
-            'prodigal',
-            '-i', assembly,
-            '-a', prots_out,
-            '-d', nucls_out,
-            '-p', 'meta',
-            '-m',
-            '-q',
-        ]
-    cmd = [str(arg) for arg in cmd]
-    logger.debug(f'cmd: {" ".join(cmd)}')
-    if parallel:
-        try:
-            returncode = subprocess.call(" ".join(cmd), shell=True)
-            tmpfpaths = glob(os.path.join(tmpdir, '*.faa'))
-            lines = ''
-            for fp in tmpfpaths:
-                with open(fp) as fh:
-                    for line in fh:
-                        lines += line
-            out = open(prots_out, 'w')
-            out.write(lines)
-            out.close()
-            tmpfpaths = glob(os.path.join(tmpdir, '*.fna'))
-            lines = ''
-            for fp in tmpfpaths:
-                with open(fp) as fh:
-                    for line in fh:
-                        lines += line
-            out = open(nucls_out, 'w')
-            out.write(lines)
-            out.close()
-        except Exception as err:
-            # COMBAK: Should probably be more descriptive as to what errors could occur here.
-            logger.exception(err)
-        finally:
-            shutil.rmtree(tmpdir)
-    else:
-        with open(os.devnull, 'w') as stdout, open(os.devnull, 'w') as stderr:
-            proc = subprocess.run(cmd, stdout=stdout, stderr=stderr)
-            returncode = proc.returncode
-    if returncode:
-        logger.warning(f'Args:{cmd} ReturnCode:{returncode}')
-        # COMBAK: Check all possible return codes for GNU parallel
+        annotate_sequential(assembly=assembly, prots_out=prots_out, nucls_out=nucls_out)
     for fp in [nucls_out, prots_out]:
-        if not os.path.exists(fp) or os.stat(fp).st_size == 0:
-            raise ChildProcessError(f'{fp} not written')
+        if not os.path.exists(fp) or os.path.getsize(fp) == 0:
+            raise ChildProcessError(f"{fp} not written")
```

Comment:
I may be wrong about this but even after `check=True` in `subprocess.run()` in  `def annotate_sequential` function, is this still needed?

Reply:
Perhaps if the prodigal run finished but no ORFs were written, we would still want to raise an error. You may be right, this may be extra unnecessary error handling.